src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java

```diff
@@ -534,71 +535,78 @@ public AzureVMDeploymentInfo call() throws Exception {
+                                                    Level.SEVERE,
+                                                    String.format("Failure to in post-provisioning for '%s'", vmName),
+                                                    e);
+
+                                            // Attempt to terminate whatever was created
+                                            AzureVMManagementServiceDelegate.terminateVirtualMachine(
+                                                    template.getAzureCloud().getServicePrincipal(), vmName,
+                                                    template.getResourceGroupName());
+                                            template.getAzureCloud().adjustVirtualMachineCount(1);
+
+                                            // Update the template status
+                                            template.handleTemplateProvisioningFailure(vmName, FailureStage.POSTPROVISIONING);
+
+                                            // Remove the node from jenkins
+                                            Jenkins.getInstance().removeNode(agent);
+                                            throw e;
                                         }
-                                        // Place node in default state, now can be
-                                        // dealt with by the cleanup task.
-                                        agent.clearCleanUpAction();
+                                        return agent;
                                     } catch (Exception e) {
-                                        LOGGER.log(
-                                                Level.SEVERE,
-                                                String.format("Failure to in post-provisioning for '%s'", vmName),
-                                                e);
-
-                                        // Attempt to terminate whatever was created
-                                        AzureVMManagementServiceDelegate.terminateVirtualMachine(
-                                                template.getAzureCloud().getServicePrincipal(), vmName,
-                                                template.getResourceGroupName());
-                                        template.getAzureCloud().adjustVirtualMachineCount(1);
-
-                                        // Update the template status
-                                        template.handleTemplateProvisioningFailure(vmName, FailureStage.POSTPROVISIONING);
-
-                                        // Remove the node from jenkins
-                                        Jenkins.getInstance().removeNode(agent);
                                         throw e;
+                                    } finally {
+                                        if ( index == numberOfNewAgents - 1) {
+                                            AzureVMAgentCleanUpTask.DeploymentLock.get(template.getResourceGroupName()).setCompleted();
```

Comment:
I'm open to a better design for decrementing the deployment refcount in a more nicer way.